src/core/lib/iomgr/tcp_client_posix.cc

```diff
@@ -62,8 +63,33 @@ struct async_connect {
   grpc_endpoint** ep;
   grpc_closure* closure;
   grpc_channel_args* channel_args;
+  int64_t connection_handle;
+  bool connect_cancelled;
 };
 
+struct ConnectionShard {
+  grpc_core::Mutex mu;
+  absl::flat_hash_map<int64_t, async_connect*> pending_connections
+      ABSL_GUARDED_BY(&mu);
+};
+
+namespace {
+
+gpr_once g_tcp_client_posix_init = GPR_ONCE_INIT;
+std::vector<ConnectionShard>* g_connection_shards = nullptr;
+std::atomic<int64_t> g_connection_id{1};
+
+void do_tcp_client_global_init(void) {
+  size_t num_shards = std::max(2 * gpr_cpu_num_cores(), 1u);
```

Comment:
I think this can be done in an implementation of `grpc_iomgr_platform_vtable.init`, or maybe `grpc_tcp_posix_init`. A new global init is redundant.

Reply:
iomgr.init implementations get invoked with every grpc_init invoked after a successful grpc_shutdown. I dont think this function implementation can be moved to a iomgr.init implementations because this needs to be executed exactly once per process. Also the global variables referenced here are static and thus cannot be referenced from another file. So I think some grpc_tcp_client_global_init function is needed

Reply:
I could call grpc_tcp_client_global_init() from tcp_posix_init(). but I think I still need to add grpc_tcp_client_global_init in iomgr_posix_cfstream.cc. Does that work ?

Reply:
I thought about commenting about this in my original review, but I decided not to bother, because the existing iomgr code already uses a lot of the `grpc_init()` infrastructure, and all of this code will go away when iomgr does.  But since AJ brought it up, here's what I was going to suggest:

```
std::vector<ConnectionShard>* GetConnectionShards() {
  static auto* shards = new std::vector<ConnectionShard>(std::max(2 * gpr_cpu_num_cores(), 1u));
  return shards;
}

std::atomic<int64_t>* GetConnectionId() {
  static auto* id = new std::atomic<int64_t>{1};
  return id;
}
```

Then change all code that accesses `g_connection_shards` and `g_connection_id` to instead use `GetConnectionShards()` and `GetConnectionId()`, respectively.

This avoids all need for explicit initialization.  However, note that when you have a function that returns a static variable like this, the compiler implicitly adds a lock around the variable's initialization, and I'm not sure if that would have any performance impact.

I think we should just do whatever's most expedient here.  We are trying to make this code go away, after all.

Reply:
Thanks for pointing this out. Since we avoid the global lock around the static variable initialization with the current approach, I am leaning towards keeping it. W.r.t performance impact, I ran the connection scaling benchmarks (maintained by @ananda1066) which establish many channels and send RPCs over them. I did not observe any regressions there.

So I'll go ahead and merge this PR for now.